scipy/sparse/linalg/isolve/lsmr.py

```diff
@@ -170,7 +173,12 @@ def lsmr(A, b, damp=0.0, atol=1e-6, btol=1e-6, conlim=1e8,
         print('atol = %8.2e                 conlim = %8.2e\n' % (atol, conlim))
         print('btol = %8.2e             maxiter = %8g\n' % (btol, maxiter))
 
-    u = b
+    if x0 is None:
+        x = zeros(n)
+        u = b
+    else:
+        x = atleast_1d(x0)
+        u = b - A.matvec(x)
```

Comment:
Like the lsqr case, this was based on the [Julia implementation](https://github.com/JuliaMath/IterativeSolvers.jl/blob/master/src/lsmr.jl) of the algorithm with initialization.  I haven't gone into detail about if or how the tolerances of the algorithm would be modified, though.